File: source/HackerCup/2022_Round_1/A/hyo_a2.cpp
#include<bits/stdc++.h>
 
using namespace std;
using ll = long long;

bool is_phase(const vector<int>& A) {
    for(auto i = 1; i < A.size(); ++i) {
        if(A.size() % i) continue;
        bool ans = true;
        for(auto j = 0; j + i < A.size(); ++i) {
            if(A[j] != A[j + i]) {
                ans = false;
                break;
            }
        }

        if(ans) return true;
    }

    return false;
}

bool is_rotate(const vector<int>& A, const vector<int>& B) {    
    auto pi = vector<int>(B.size());
    for(auto i = 1, j = 0; i < B.size(); ++i)
    {
        while(j && B[i] != B[j]) j = pi[j-1];
        if(B[i] == B[j]) pi[i] = ++j;
    }

    vector<int> AA = A;
    AA.insert(AA.end(), A.begin(), A.end());
    for(auto i = 0, j = 0; i < AA.size(); ++i)
    {
        while(j && AA[i] != B[j]) j = pi[j-1];
        if(AA[i] == B[j]) {
            if(j == B.size() - 1) return true;
            else j++;
        }
    }

    return false;
}

void solve(int TestCase) {
    int n, k;
    cin >> n >> k;

    vector<int> A(n), B(n);
    for(auto& x : A) cin >> x;
    for(auto& x : B) cin >> x;

    bool ans = true;
    if(k == 0) {
        ans = A == B;
    } else if (k == 1) {
        ans = A == B && is_phase(A) || A != B && is_rotate(A, B);
    } else if (n == 2) {
        if(k % 2 == 0) ans = A == B;
        else ans = A == B && is_phase(A) || A != B && is_rotate(A, B);
    } else {
        ans = is_rotate(A, B);
    }

    if(ans) cout << "YES\n";
    else cout << "NO\n";
}
 
int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
 
    int t = 1;
    cin >> t;
    for(auto i = 1; i <= t; ++i) {
        cout << "Case #"<< i << ": ";
        solve(i);
    }
}

File: source/HackerCup/2022_Round_1/A/hyo_a2.rs
use std::io::Read;

pub fn get_pi<T>(s: &Vec<T>) -> Vec<usize>
where
    T: std::cmp::Eq,
{
    let n = s.len();
    let mut j = 0;
    let mut pi = vec![0; n];

    for i in 1..n {
        while j > 0 && s[i] != s[j] {
            j = pi[j - 1];
        }

        if s[i] == s[j] {
            j += 1;
            pi[i] = j;
        }
    }

    pi
}

pub fn kmp<T>(s: &Vec<T>, p: &Vec<T>) -> Vec<usize>
where
    T: std::cmp::Eq,
{
    let mut ret = Vec::new();
    let n = s.len();
    let m = p.len();
    let mut j = 0;
    let pi = get_pi::<T>(p);
    for i in 0..n {
        while j > 0 && s[i] != p[j] {
            j = pi[j - 1];
        }
        if s[i] == p[j] {
            if j == m - 1 {
                ret.push(i + 1 - m);
                j = pi[j];
            } else {
                j += 1;
            }
        }
    }

    ret
}

fn is_rotate(a: &Vec<usize>, b: &Vec<usize>) -> bool {
    let mut a = a.clone();
    a.extend(a.clone().into_iter());

    !kmp::<usize>(&a, &b).is_empty()
}

fn is_phase(a: &Vec<usize>) -> bool {
    let n = a.len();
    for len in 1..n {
        if n % len != 0 {
            continue;
        }

        let mut ans = true;
        for i in len..n {
            if a[i - len] != a[i] {
                ans = false;
                break;
            }
        }

        if ans {
            return true;
        }
    }

    false
}

fn main() {
    let mut stdin = std::io::stdin();
    let mut buf = String::new();

    stdin.read_to_string(&mut buf).unwrap();

    let mut input = buf.split_ascii_whitespace();

    let test_case = input.next().unwrap().parse::<usize>().unwrap();
    for tc in 1..=test_case {
        let n = input.next().unwrap().parse::<usize>().unwrap();
        let k = input.next().unwrap().parse::<usize>().unwrap();

        let a = (0..n)
            .map(|_| input.next().unwrap().parse::<usize>().unwrap())
            .collect::<Vec<_>>();

        let b = (0..n)
            .map(|_| input.next().unwrap().parse::<usize>().unwrap())
            .collect::<Vec<_>>();

        let rotate = is_rotate(&a, &b);
        let ans = if k == 0 {
            a == b
        } else if k == 1 {
            if a == b {
                is_phase(&a)
            } else {
                rotate
            }
        } else if n == 2 {
            if a[0] == a[1] {
                true
            } else if k % 2 == 0 {
                a == b
            } else {
                a != b
            }
        } else {
            rotate
        };

        let ans = if ans { "YES" } else { "NO" };
        println!("Case #{tc}: {ans}");
    }
}
